File: tomo_core/src/tomo_core/telegram.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol

from .models import MAX_REPLY_CONTEXT_TEXT_CHARS, MessageAttachment, OutboundBubble, ReplyContext


def reply_context_from_message(message: dict, parent_chat_id: object) -> ReplyContext | None:
    reply = message.get("reply_to_message")
    if not isinstance(reply, dict) or reply.get("message_id") is None:
        return None
    reply_chat = reply.get("chat")
    if not isinstance(reply_chat, dict) or str(reply_chat.get("id")) != str(parent_chat_id):
        return None
    sender = reply.get("from")
    author_role = "assistant" if isinstance(sender, dict) and sender.get("is_bot") else "user" if isinstance(sender, dict) else "unknown"
    text = reply.get("text") if isinstance(reply.get("text"), str) else reply.get("caption") if isinstance(reply.get("caption"), str) else None
    truncated = bool(text and len(text) > MAX_REPLY_CONTEXT_TEXT_CHARS)
    if text is not None:
        text = text[:MAX_REPLY_CONTEXT_TEXT_CHARS]
    attachments = photo_attachments_from_message(reply)
    availability = "available" if text is not None or attachments else "unavailable"
    return ReplyContext(
        message_id=str(reply["message_id"]), author_role=author_role, text=text,
        timestamp=str(reply["date"]) if reply.get("date") is not None else None,
        attachments=attachments, availability=availability, truncated=truncated,
    )


def photo_attachments_from_message(message: dict) -> tuple[MessageAttachment, ...]:
    photos = message.get("photo")
    if not isinstance(photos, list):
        return ()
    photo = max((item for item in photos if isinstance(item, dict) and isinstance(item.get("file_id"), str)), key=lambda item: int(item.get("width") or 0) * int(item.get("height") or 0), default=None)
    if photo is None:
        return ()
    return (MessageAttachment("image", file_id=photo["file_id"], mime_type="image/jpeg", metadata={key: photo[key] for key in ("width", "height", "file_size", "file_unique_id") if key in photo}),)


@dataclass(frozen=True)
class TelegramSendReceipt:
    message_id: str


class TelegramClient(Protocol):
    def send_typing(self, actor_id: str) -> None:
        ...

    def send_message(self, actor_id: str, text: str, reply_to_message_id: str | None = None) -> TelegramSendReceipt:
        ...

    def set_message_reaction(self, actor_id: str, message_id: str, emoji: str) -> None:
        ...


@dataclass
class TelegramDeliverySink:
    client: TelegramClient

    def start_typing(self, actor_id: str) -> None:
        self.client.send_typing(actor_id)

    def send_bubbles(self, actor_id: str, bubbles: list[OutboundBubble]) -> None:
        for bubble in bubbles:
            self.client.send_message(
                actor_id=actor_id,
                text=bubble.text,
                reply_to_message_id=bubble.reply_to_message_id,
            )

    def react_to_message(self, actor_id: str, message_id: str, emoji: str) -> None:
        self.client.set_message_reaction(actor_id, message_id, emoji)


@dataclass
class FakeTelegramClient:
    typing_actor_ids: list[str] = field(default_factory=list)
    sent_messages: list[dict[str, str | None]] = field(default_factory=list)
    reactions: list[dict[str, str]] = field(default_factory=list)

    def send_typing(self, actor_id: str) -> None:
        self.typing_actor_ids.append(actor_id)

    def send_message(self, actor_id: str, text: str, reply_to_message_id: str | None = None) -> TelegramSendReceipt:
        self.sent_messages.append(
            {"actor_id": actor_id, "text": text, "reply_to_message_id": reply_to_message_id}
        )
        return TelegramSendReceipt(str(len(self.sent_messages)))

    def set_message_reaction(self, actor_id: str, message_id: str, emoji: str) -> None:
        self.reactions.append({"actor_id": actor_id, "message_id": message_id, "emoji": emoji})

File: cloudflare/src/telegram.ts
export interface CompactUpdate {
  updateId: number;
  chatId: string;
  actorId: string;
  messageId: string | null;
  text: string;
  date: number;
  photoFileId: string | null;
  photoMeta: Record<string, string | number>;
  replyTo: {
    messageId: string;
    authorRole: "user" | "assistant" | "unknown";
    text: string | null;
  } | null;
  raw: Record<string, unknown>;
}

export function compactPrivateMessage(update: Record<string, unknown>): CompactUpdate | null {
  const updateId = update.update_id;
  const message = update.message;
  if (typeof updateId !== "number" || message === null || typeof message !== "object") return null;
  const msg = message as Record<string, unknown>;
  const chat = msg.chat;
  const from = msg.from;
  if (chat === null || typeof chat !== "object" || from === null || typeof from !== "object") return null;
  const chatObj = chat as Record<string, unknown>;
  const fromObj = from as Record<string, unknown>;
  if (chatObj.type !== "private") return null;
  const chatId = String(chatObj.id ?? "");
  const actorId = String(fromObj.id ?? "");
  if (!chatId || !actorId) return null;
  const photos = Array.isArray(msg.photo) ? msg.photo : [];
  let photoFileId: string | null = null;
  let photoMeta: Record<string, string | number> = {};
  let best = -1;
  for (const item of photos) {
    if (item === null || typeof item !== "object") continue;
    const photo = item as Record<string, unknown>;
    const area = Number(photo.width || 0) * Number(photo.height || 0);
    if (typeof photo.file_id === "string" && area >= best) {
      best = area;
      photoFileId = photo.file_id;
      photoMeta = {};
      for (const key of ["width", "height", "file_size", "file_unique_id"] as const) {
        if (key in photo && (typeof photo[key] === "number" || typeof photo[key] === "string")) {
          photoMeta[key] = photo[key] as number | string;
        }
      }
    }
  }
  const text = typeof msg.text === "string" ? msg.text : typeof msg.caption === "string" ? msg.caption : "";
  if (!text && !photoFileId) return null;
  let replyTo: CompactUpdate["replyTo"] = null;
  const reply = msg.reply_to_message;
  if (reply !== null && typeof reply === "object") {
    const parent = reply as Record<string, unknown>;
    if (parent.message_id !== undefined) {
      const sender = parent.from;
      let authorRole: "user" | "assistant" | "unknown" = "unknown";
      if (sender !== null && typeof sender === "object") {
        authorRole = (sender as { is_bot?: boolean }).is_bot ? "assistant" : "user";
      }
      const replyText =
        typeof parent.text === "string" ? parent.text : typeof parent.caption === "string" ? parent.caption : null;
      replyTo = { messageId: String(parent.message_id), authorRole, text: replyText };
    }
  }
  return {
    updateId,
    chatId,
    actorId,
    messageId: msg.message_id === undefined ? null : String(msg.message_id),
    text,
    date: typeof msg.date === "number" ? msg.date : 0,
    photoFileId,
    photoMeta,
    replyTo,
    raw: update,
  };
}

export class TelegramApi {
  private token: string;

  constructor(token: string) {
    this.token = token;
  }

  async sendMessage(chatId: string, text: string, replyTo?: string | null): Promise<{ messageId: string } | { uncertain: boolean }> {
    const body: Record<string, unknown> = { chat_id: chatId, text };
    if (replyTo) body.reply_to_message_id = Number(replyTo);
    try {
      const response = await fetch(`https://api.telegram.org/bot${this.token}/sendMessage`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(body),
      });
      if (!response.ok) return { uncertain: false };
      const payload = (await response.json()) as { ok?: boolean; result?: { message_id?: number } };
      if (!payload.ok || payload.result?.message_id === undefined) return { uncertain: true };
      return { messageId: String(payload.result.message_id) };
    } catch {
      return { uncertain: true };
    }
  }

  async sendTyping(chatId: string): Promise<void> {
    try {
      await fetch(`https://api.telegram.org/bot${this.token}/sendChatAction`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({ chat_id: chatId, action: "typing" }),
      });
    } catch {
      // typing is best-effort
    }
  }

  async react(
    chatId: string,
    messageId: string,
    emoji: string,
  ): Promise<{ ok: true } | { ok: false; code: string }> {
    const numericMessageId = Number(messageId);
    if (!chatId || !Number.isSafeInteger(numericMessageId) || numericMessageId <= 0) {
      return { ok: false, code: "invalid_target" };
    }
    try {
      const response = await fetch(`https://api.telegram.org/bot${this.token}/setMessageReaction`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          chat_id: chatId,
          message_id: numericMessageId,
          reaction: [{ type: "emoji", emoji }],
          is_big: false,
        }),
      });
      let payload: { ok?: boolean; result?: boolean; error_code?: number };
      try {
        payload = (await response.json()) as { ok?: boolean; result?: boolean; error_code?: number };
      } catch {
        return { ok: false, code: response.ok ? "invalid_response" : `http_${response.status}` };
      }
      if (response.ok && payload.ok === true && payload.result === true) return { ok: true };
      if (typeof payload.error_code === "number") return { ok: false, code: `telegram_${payload.error_code}` };
      return { ok: false, code: response.ok ? "telegram_rejected" : `http_${response.status}` };
    } catch {
      return { ok: false, code: "transport" };
    }
  }

  async fetchFile(fileId: string): Promise<{ bytes: Uint8Array; mime: string } | null> {
    const meta = await fetch(`https://api.telegram.org/bot${this.token}/getFile`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ file_id: fileId }),
    });
    if (!meta.ok) return null;
    const payload = (await meta.json()) as { ok?: boolean; result?: { file_path?: string } };
    if (!payload.ok || !payload.result?.file_path) return null;
    const file = await fetch(`https://api.telegram.org/file/bot${this.token}/${payload.result.file_path}`);
    if (!file.ok) return null;
    const bytes = new Uint8Array(await file.arrayBuffer());
    if (bytes.byteLength > 10 * 1024 * 1024) return null;
    const contentType = file.headers.get("content-type") || "";
    const mime = contentType.startsWith("image/") ? contentType : photoMime(payload.result.file_path);
    return { bytes, mime };
  }
}

function photoMime(filePath: string): string {
  if (/\.png$/i.test(filePath)) return "image/png";
  if (/\.webp$/i.test(filePath)) return "image/webp";
  return "image/jpeg";
}

export async function equalSecret(left: string, right: string): Promise<boolean> {
  const encoder = new TextEncoder();
  const [a, b] = await Promise.all([
    crypto.subtle.digest("SHA-256", encoder.encode(left)),
    crypto.subtle.digest("SHA-256", encoder.encode(right)),
  ]);
  const leftBytes = new Uint8Array(a);
  const rightBytes = new Uint8Array(b);
  let diff = 0;
  for (let i = 0; i < leftBytes.length; i += 1) diff |= leftBytes[i] ^ rightBytes[i];
  return diff === 0;
}
